SteamKit2/SteamKit2/Steam/Handlers/SteamMatchmaking/Lobby.cs

```diff
@@ -0,0 +1,402 @@
+
+            /// <summary>
+            /// The maximum number of members that can occupy the lobby.
+            /// </summary>
+            public int MaxMembers { get; }
+
+            /// <summary>
+            /// The number of members that are currently occupying the lobby.
+            /// </summary>
+            public int NumMembers { get; }
+
+            /// <summary>
+            /// A list of lobby members. This will only be populated for the user's current lobby.
+            /// </summary>
+            public IReadOnlyList<Member> Members { get; }
+
+            /// <summary>
+            /// The distance of the lobby.
+            /// </summary>
+            public float? Distance { get; }
+
+            /// <summary>
+            /// The weight of the lobby.
+            /// </summary>
+            public long? Weight { get; }
+
+            static readonly IReadOnlyDictionary<string, string> EmptyMetadata =
+                new ReadOnlyDictionary<string, string>( new Dictionary<string, string>() );
+
+            static readonly IReadOnlyList<Member> EmptyMembers = new List<Member>().AsReadOnly();
+
+            internal Lobby( SteamID steamId, ELobbyType lobbyType, int lobbyFlags, SteamID ownerSteamId, IReadOnlyDictionary<string, string> metadata,
+                int maxMembers, int numMembers, IReadOnlyList<Member> members, float? distance, long? weight )
+            {
+                SteamID = steamId;
+                LobbyType = lobbyType;
+                LobbyFlags = lobbyFlags;
+                OwnerSteamID = ownerSteamId;
+                Metadata = metadata ?? EmptyMetadata;
+                MaxMembers = maxMembers;
+                NumMembers = numMembers;
+                Members = members ?? EmptyMembers;
+                Distance = distance;
+                Weight = weight;
+            }
+
+            internal static byte[] EncodeMetadata( IReadOnlyDictionary<string, string> metadata )
+            {
+                if ( metadata == null )
+                {
+                    metadata = EmptyMetadata;
+                }
+
+                var keyValue = new KeyValue( "" );
+
+                foreach ( var entry in metadata )
+                {
+                    keyValue[ entry.Key ] = new KeyValue( null, entry.Value );
+                }
+
+                using ( var ms = new MemoryStream() )
+                {
+                    keyValue.SaveToStream( ms, true );
+
+                    var buffer = new byte[ ms.Length ];
+                    ms.Seek( 0, SeekOrigin.Begin );
+                    ms.Read( buffer, 0, buffer.Length );
+                    return buffer;
+                }
+            }
+
+            internal static ReadOnlyDictionary<string, string> DecodeMetadata( byte[] buffer )
+            {
+                var metadata = new Dictionary<string, string>();
+
+                if ( buffer.Length > 0 )
+                {
```

Comment:
Can we move the allocation of the dictionary here, and re-use `EmptyMetadata` if the length is zero?

Reply:
Sure... and after my previous comment; here I am writing code with worse cyclomatic complexity :wink: